server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDeciders.java

```diff
@@ -34,279 +37,202 @@ public class AllocationDeciders {
+                node,
+                decider,
+                decision
+            )
+        );
+    }
+
+    private Decision withDeciders(
+        RoutingAllocation allocation,
+        Function<AllocationDecider, Decision> deciderAction,
+        BiFunction<String, Decision, String> logMessageCreator
+    ) {
+        return withDeciders(allocation.getDebugMode(), deciderAction, logMessageCreator);
+    }
+
+    private Decision withDecidersCheckingShardIgnoredNodes(
+        RoutingAllocation allocation,
+        ShardRouting shardRouting,
+        RoutingNode node,
+        Function<AllocationDecider, Decision> deciderAction,
+        BiFunction<String, Decision, String> logMessageCreator
+    ) {
         if (allocation.shouldIgnoreShardForNode(shardRouting.shardId(), node.nodeId())) {
-            return NO_IGNORING_SHARD_FOR_NODE;
-        }
-        Decision.Multi ret = new Decision.Multi();
-        for (AllocationDecider allocationDecider : allocations) {
-            Decision decision = allocationDecider.canAllocateReplicaWhenThereIsRetentionLease(shardRouting, node, allocation);
-            // short track if a NO is returned.
-            if (decision.type() == Decision.Type.NO) {
-                if (allocation.debugDecision() == false) {
-                    return Decision.NO;
-                } else {
-                    ret.add(decision);
-                }
-            } else {
-                addDecision(ret, decision, allocation);
+            if (logger.isTraceEnabled()) {
+                logger.trace(() -> logMessageCreator.apply(AllocationDeciders.class.getSimpleName(), NO_IGNORING_SHARD_FOR_NODE));
             }
+            return NO_IGNORING_SHARD_FOR_NODE;
         }
-        return ret;
+        return withDeciders(allocation.getDebugMode(), deciderAction, logMessageCreator);
     }
 
-    private static void addDecision(Decision.Multi ret, Decision decision, RoutingAllocation allocation) {
-        // We never add ALWAYS decisions and only add YES decisions when requested by debug mode (since Multi default is YES).
-        if (decision != Decision.ALWAYS
-            && (allocation.getDebugMode() == RoutingAllocation.DebugMode.ON || decision.type() != Decision.Type.YES)) {
-            ret.add(decision);
+    private Decision withDeciders(
+        RoutingAllocation.DebugMode debugMode,
+        Function<AllocationDecider, Decision> deciderAction,
+        BiFunction<String, Decision, String> logMessageCreator
+    ) {
+        if (debugMode == RoutingAllocation.DebugMode.OFF) {
+            var result = Decision.YES;
+            for (AllocationDecider decider : deciders) {
+                var decision = deciderAction.apply(decider);
+                if (decision.type() == Decision.Type.NO) {
+                    if (logger.isTraceEnabled()) {
+                        logger.trace(() -> logMessageCreator.apply(decider.getClass().getSimpleName(), decision));
+                    }
```

Comment:
I wonder if there is a value in logging throttling decisions as well as they could prevent allocations

Reply:
I'm ok with this (for now at least)